File: drills/ts/solutions/nc106_number_of_islands__union_find.ts
export function numIslands(grid: string[][]): number {
  const land = new Set<string>();
  grid.forEach((row, r) => row.forEach((value, c) => value === "1" && land.add(`${r},${c}`)));

  // Union-find instead of flood fill: every square starts as its own island and
  // each adjacency merges two. Only right and down are needed -- every pair of
  // neighbours is reached once that way -- and the answer is how many roots are
  // left. This is the version that keeps working when the grid arrives one
  // square at a time and the count has to be reported after each.
  const parents = new Map<string, string>();
  for (const at of land) parents.set(at, at);

  const find = (at: string): string => {
    while (parents.get(at) !== at) {
      parents.set(at, parents.get(parents.get(at)!)!);
      at = parents.get(at)!;
    }
    return at;
  };

  for (const at of land) {
    const [r, c] = at.split(",").map(Number);
    for (const other of [`${r + 1},${c}`, `${r},${c + 1}`]) {
      if (!land.has(other)) continue;
      const a = find(at);
      const b = find(other);
      if (a !== b) parents.set(a, b);
    }
  }

  return new Set([...land].map(find)).size;
}

File: drills/go/solutions/nc106_number_of_islands__union_find.go
package main

func numIslands(grid [][]byte) int {
	if len(grid) == 0 {
		return 0
	}
	rows, cols := len(grid), len(grid[0])
	// Every land cell is its own set; union it with land to the right and
	// below. The islands are the sets left standing.
	parent := make([]int, rows*cols)
	for i := range parent {
		parent[i] = i
	}
	var find func(i int) int
	find = func(i int) int {
		for parent[i] != i {
			parent[i] = parent[parent[i]]
			i = parent[i]
		}
		return i
	}
	count := 0
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == '1' {
				count++
			}
		}
	}
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] != '1' {
				continue
			}
			for _, d := range [][2]int{{1, 0}, {0, 1}} {
				nr, nc := r+d[0], c+d[1]
				if nr < rows && nc < cols && grid[nr][nc] == '1' {
					a, b := find(r*cols+c), find(nr*cols+nc)
					if a != b {
						parent[a] = b
						count--
					}
				}
			}
		}
	}
	return count
}
